### ISSUE TITLE
[BUG] XPS DrawBitMap is clipping?

### ISSUE THREAD

**Issue by Jonesie** [mono/SkiaSharp#1728]
**Description**

Im using SkiaSharp to create an XPS (and/or PDF) document that includes images for printing. Images in PDF are correctly stretched to fit the required rect, but on XPS they seem to be clipped. My code and sample images are below.

Is this a bug or am I doing something wrong? I would have thought that the code for PDF should be identical for XPS but maybe there's extra trick or hack I need to do?

**Code**

```
public void ToDocument(string inputPath, bool pdf = false)
        {

            SKDocument document = null;
            var outputPath = Path.Combine(Path.GetDirectoryName(inputPath), Path.GetFileNameWithoutExtension(inputPath)) + (pdf ? ".pdf" : ".xps");

            using var input = File.OpenRead(inputPath);
            using var inputStream = new SKManagedStream(input);
            using var original = SKBitmap.Decode(inputStream);

            // page a little bigger than source image
            var pageRect = new SKRect(0, 0, original.Width + 20, original.Height + 100);

            var renderRect = new SKRect(0, 0,
                (float)original.Width, (float)original.Height);

            using var outputStream = File.Create(outputPath);
            using var managedStream = new SKManagedWStream(outputStream);

            // null checking removed for readability

            using var canvas = document.BeginPage(pageRect.Width, pageRect.Height);
            canvas.Clear();

            using var p = new SKPaint();

            SKRect picRect;
            float xFactor = 0.25F;
            float yFactor = 0.25F;
            if (pdf)
            {
                // stretch renderRect for pdf
                picRect = new SKRect(renderRect.Left, renderRect.Top, renderRect.Left + renderRect.Width * xFactor, renderRect.Top + renderRect.Height * yFactor);
            }
            else
            {
                // stretch renderRect for xps
                // xFactor = 1.25F;
                // yFactor = 1.25F;
                picRect = new SKRect(renderRect.Left, renderRect.Top, renderRect.Left + renderRect.Width * xFactor, renderRect.Top + renderRect.Height * yFactor);
            }
            canvas.DrawBitmap(original, original.Info.Rect, picRect, p);

            p.StrokeWidth = 2;
            p.Color = SKColors.Red;
            p.Style = SKPaintStyle.Stroke;

            // rect same size as original bitmap
            canvas.DrawRect(renderRect, p);

            p.Color = SKColors.Blue;
            // rect same size as stretched bitmap
            canvas.DrawRect(picRect, p);

            using var tp = new SKPaint();
            tp.TextSize = 10.0f;
            tp.IsAntialias = true;
            tp.IsStroke = false;
            //canvas.DrawText($"Page: {pageRect.Height}w {pageRect.Width}h", 10F, pageRect.Bottom - 55F, tp);
            tp.Color = SKColors.Red;
            canvas.DrawText($"Source: {original.Width}w {original.Height}h", 10F, pageRect.Bottom - 40F, tp);
            tp.Color = SKColors.Blue;
            canvas.DrawText($"Dest: {picRect.Width}w {picRect.Height}h", 10F, pageRect.Bottom - 25F, tp);
            canvas.DrawText($"Stretched: {xFactor}w {yFactor}h", 10F, pageRect.Bottom - 10F, tp);


            // bitmap should fit to the background rect but fails with xps
            //canvas.DrawBitmap(original, renderRect, p);

            document.Close();
        }
```

**Expected Behavior**

**Actual Behavior**

Some images are clipped - it seems like small ones are ok.  Maybe there is a buffer size issue or ... ?

**Basic Information**

- Version with issue:  2.80.2
- Last known good version:  ??
- IDE:   Visual Studio, Visual Studio Code 
- Platform Target Frameworks: 
  - Linux:  Ubuntu 16.04 (AWS Linux)
  - Windows Classic:  10
- Target Devices:   
  - n/a - printing the xps
  
<details>
  <summary>Detailed IDE/OS information (click to expand)</summary>
  
```

PASTE ANY DETAILED VERSION INFO HERE

```
</details>

**Screenshots**
Original image:
![Original](https://user-images.githubusercontent.com/3420249/123340079-c1086000-d59f-11eb-9ba1-c108e00f4e6c.png)

PDF: 
![PDF](https://user-images.githubusercontent.com/3420249/123340085-c4035080-d59f-11eb-95fc-0d671bc63bad.png)

XPS:
![XPS](https://user-images.githubusercontent.com/3420249/123340093-c5cd1400-d59f-11eb-8b18-d5e03ca25409.png)

**Comment by AndreaLabeljoy** [comment]
Hi @Jonesie, ran into this exact issue today on version 2.88.8.
Did you ever find a way to  make it work?

**Comment by oatkins** [comment]
It turns out that this is a bug in the Google Skia code. It's even commented 😆:
* https://github.com/google/skia/blob/c30ba7bb38d6f368631db17d76628e89b0fe5004/src/xps/SkXPSDevice.cpp#L685
* https://github.com/google/skia/blob/c30ba7bb38d6f368631db17d76628e89b0fe5004/src/xps/SkXPSDevice.cpp#L796

![image](https://github.com/user-attachments/assets/c58f2786-a4f0-49de-b970-bdf8f1f99265)

The code alludes to `SK_ScalarMax`, which is large, but for some reason they ended up using 1000 instead. The result is that all images get cropped to a 1000 pixel square centred on the origin. Generally, that means that images get cropped to 500 pixels high and 500 pixels wide.

It looks as though Windows does *not* consider the `VisualBrush.Viewport` to be a clipping rectangle, but WPF does. I had, but now can't find, a reference to that in the WPF docs; the XPS spec is vague about it, and doesn't mention clipping at all. The Windows XPS viewer (and probably the printer drivers) render content outside of the viewport, which is probably why the bug has gone unnoticed for so long. So, perhaps the WPF behaviour is wrong.

In any case, my workaround (in my own code) is going to be adding some code to find the `Viewbox` and `Viewport` values in XAML:
![image](https://github.com/user-attachments/assets/05df24d5-9d57-4803-a6aa-d9866d8d3287)

and replace them with a much, much larger box.

In short, I don't think there's much that can be done in this repo, but there is a potential workaround for those hitting this problem.

**Comment by Jonesie** [comment]
Thats good to know.  Thank you !!!

**Comment by AndreaLabeljoy** [comment]
Thanks @oatkins for the info.
I've created a bug report on the Skia group referring them to this post, let's see if someone addresses it.
[https://issues.skia.org/issues/360641492](url)
